source/blog/029-node-fibers-discontinued.md

```diff
@@ -0,0 +1,160 @@
+allocate callbacks, store them somewhere, and take a trip back to the event loop
+before invoking those callbacks, and that all takes time. In fact, it takes
+enough time that the asynchronous `render()` in Dart Sass tends to be 2-3x
+slower than `renderSync()`.
+
+Enter fibers. Fibers are a very cool concept, available in languages like Ruby
+and C++, that give the programmer more control over asynchronous functions. They
+can even allow a chunk of synchronous code (such as the Sass compiler) to call
+asynchronous callbacks (such as the webpack plugin). The `node-fibers` package
+did some arcane magick with the V8 virtual machine to implement Fibers in
+JavaScript, which allowed Dart Sass to use the fast synchronous code to
+implement the asynchronous `render()` API. And for a time, it was great.
+
+### The Death of Fibers
+
+Unfortunately, the arcane magick that `node-fibers` used involved accessing some
+parts of V8 that were not officially part of its public API. There was no
+guarantee that the interfaces they were using would stay the same from release
+to release, and indeed they tended to change fairly regularly. For a long time,
+those changes were small enough that it was possible to release a new version of
+`node-fibers` that supported them, but with Node.js 16 the luck ran out.
+
+The latest version of V8 involves some major overhauls to its internals. These
+will eventually allow it to implement some cool improvements, so its hard to
+begrudge, but a side effect is that the APIs `node-fibers` was using are
+completely gone without an obvious replacement. This is no one's fault: since
+those interfaces weren't part of V8's public API, they were under no obligation
+to keep them stable. Sometimes in software that's just the way things go.
+
+## Reclaiming Performance
+
+There are a few options for getting back the performance that's lost by no
+longer being able to pass `node-fibers` to `sass.render()`. In order from
+nearest to longest term:
+
+### Avoid Asynchronous Plugins
+
+This is something you can do today. If it's at all possible to make the plugins
+you pass in to Sass synchronous, you can use the `renderSync()` method which
+doesn't need fibers to go fast. This may require rewriting some existing
+plugins, but it will pay dividends immediately.
+
+### Embedded Dart Sass
+
+While it's not ready for prime-time yet, the Sass team is working on a project
+called "embedded Dart Sass". This involves running Dart Sass as a *subprocess*,
+rather than a library, and communicating with it using a special protocol. This
+provides several important improvements over the existing alternatives:
+
+* Unlike running `sass` from the command line, this will still work with plugins
+  like the webpack importer. In fact, we plan to match the existing JavaScript
+  API as closely as possible. This will probably run asynchronous plugins *even
+  faster* than synchronous ones.
```

Comment:
Not necessary to mention in the blog post, but I'm curious why async plugins would be faster than sync ones here (unless this is supposed to mean that async plugins on embedded Dart Sass are faster than synchronous plugins on compiled-to-JS Dart Sass?)

Reply:
They're faster because it's actually possible to run in parallel with other processing the way evented programming is *supposed* to work. In JS Dart Sass, any parallelism gains are overshadowed by the overhead from asynchrony, but Embedded Dart Sass actually does all the hard work of compilation out-of-process, so Node.js can continue to do productive work in the meantime. If the entire app is single-threaded it'll still be slower, but in a webpack-style environment where multiple builds are running at once it should be a win.